Feature/allow master control

pmulh opened the pull request:
#### Description
Two separate changes:
- Enabling "what song is this?"/"What are we listening to?" functionality, that was mostly already implemented, but commented out...? It seems to all work well (#7 )
- Adding option to allow mycroft to control playback on other spotify devices, if allow_master_control is set to true in settings.json (e.g. 'play the artist the beatles' will start playback on the primary device; 'transfer spotify to <name of another spotify device>' will transfer playback, then 'pause'/'next track' etc will control playback on the device that playback was transferred to).

#### Type of PR
If your PR fits more than one category, there is a high chance you should submit more than one PR. Please consider this carefully before opening the PR.
_Either delete those that do not apply, or add an x between the square brackets like so: `- [x]`_
- [ ] Bugfix
- [X] Feature implementation
- [ ] Refactor of code (without functional changes)
- [ ] Documentation improvements
- [ ] Test improvements

#### Testing
These changes have only been tested on picroft; if anyone can test them on a Mark1 or Mark2, that would be good.

#### Documentation
Does documentation for this already exist? Are there docstrings on all the public methods you added or modified? Have these been updated?

pmulh committed:
Add now playing info, and ability to control playback started elsewhere

pmulh committed:
Merge branch 'forslund:21.02' into feature/allow_master_control

forslund commented:
Hi, sorry for the late reply.

I need to dig into and try it out properly but in general I think it looks good.

The thing I'm wondering about is the queries for asking about what's playing. It was my intention that the queries would only be available if music was actually playing...I need to check if that was working as intended in the first place.

forslund commented:
I've just refreshed my memory and the "what song is this / what are we listening to" intents are currently enabled when the spotify skill is playing. Maybe you can change it to be default on when "master control" is enabled, and keep it as it is (enabling on play, disabling on stop)

pmulh commented:
Sorry, I forgot to reply to your previous comment.
Thanks for the clarification of how the intents are currently enabled/disabled, I hadn't appreciated that. I'll have a look at changing the default when "master control" is enabled like you suggested

pmulh committed:
Merge branch 'forslund:21.02' into feature/allow_master_control

pmulh committed:
Updates to how "what song is this"-type intents work when allow_master_control is true

Some previous changes have been reverted to reduce changes when allow_master_control is false.

One change that applies regardless of the value of allow_master_control, though, is the addition of a call to enable_playing_intents() in the resume() function, as it seemed like the current functionality would mean that the "what song is this"-type intents would be disabled when the music was paused, and not re-enabled once the music is resumed (which is presumably the desired behaviour).

pmulh commented:
I pushed another commit there to address the issue you pointed out about how the "what song is this" etc intents are enabled/disabled.
Now it should be that when allow_master_control is true, they aren't disabled when music stops playing, but are when allow_master_control is false and music stops (i.e. no change).

When I was testing it I did notice that currently (so with allow_master_control == False), those intents are disabled if the music is paused, but aren't re-enabled when the music is resumed. So I added a call to enable_playing_intents() in the resume() function to address that (assuming that these intents should always be enabled when music is playing).

The changes in song_info(), artist_info(), album_info() are to prevent mycroft saying something like "we're listening to None by None" if allow_master_control is True but nothing is currently playing (it will say the NothingPlaying dialog instead).

forslund commented on __init__.py:
```diff
@@ -809,10 +811,13 @@ def enable_playing_intents(self):
-        self.disable_intent('StopMusic.intent')
+        # If allow_master_control is false, disable these intents (called
+        # when music stops playing)
+        if not self.allow_master_control:
```
I think it would be more expressive if this check is done where the `disable_playing_intents()` method is called so the code at those sites would be

```python
if not self.allow_master_control:
    self.disable_intent()
```

pmulh replied:
Cool, yeah that makes sense. I'll make that change

forslund commented on __init__.py:
```diff
@@ -125,6 +125,7 @@ def update_librespot():
         print('Librespot Update failed, {}'.format(repr(e)))
 
 
+@intent_handler(['WhatSong.intent', 'WhatAlbum.intent', 'WhatArtist.intent'])
```
Does this work?

This function has no access to the bus and when called the status will be a Message and not a status dict.

pmulh replied:
Good point, thanks. I think I originally thought this would be required for the intents, but looking at it again I think it's actually doing nothing. I've removed it and it seems to have had no effect; I'll do a few more tests later to double check, but I think this can be removed.

forslund reviewed (commented):
Sorry that it has taken me a long time to get back to you on this. I still need to do more trials but I'll post the few comments / questions I had right now.

pmulh committed:
Updates to handling of Now Playing intents, based on PR comments

pmulh committed:
Removing unnecessary logging call

pmulh committed:
Tidy-up of comments

pmulh commented:
> Sorry that it has taken me a long time to get back to you on this. I still need to do more trials but I'll post the few comments / questions I had right now.

No worries, thanks for those comments. I've addressed those in the latest commits.

Minor additional point: one thing I noticed when doing some testing was that currently (in this branch and in 21.02 I believe) if you do something like "transfer spotify to Device2", while the playback control intents (pause, next track etc) will no longer work, the "what song is this" intents will still work, since the self.spotify.is_playing() check on line 383 checks for playback on any device. Personally I don't think this is an issue, but just mentioning it in case this isn't the intended functionality.

pmulh committed:
Reworking of how functionality for allowing control of other devices works

pmulh commented:
Sorry to completely rework the changes mid-pull request, but I was having another look at the code and doing some more testing after your comments last week, and noticed a couple of areas that could be improved. I think the most recent commit is a much more robust set of changes than the previous commit.

The new changes use the current_playback() spotipy function to check what the current playing or most recently played (with a cutoff of about 10 minutes after playback is paused) device is, which avoids weird issues where the skill is trying to pause/resume/skip track etc on the wrong device. This only applies when that allow_master_control config is set to True.

The changes when allow_master_control are False should be very minimal - a minor bugfix or two, and I added return statements to prev_track() so that it matches the structure of next_track().

Also I realise that this is probably not a particularly high priority enhancement for the skill, so I understand if you don't want to incorporate these changes right now. They're very useful for my particular setup, but may not be as useful to others.

forslund commented on __init__.py:
```diff
@@ -1274,6 +1354,12 @@ def do_stop(self):
+        # not work, but pause should work regardless of which device is playing
+        if self.spotify and self.allow_master_control:
+            self.pause()
+            return True
```
Should this check if a unit is playing before performing the pause and returning true?

In this case the skill will always consume the stop event if it's configured and for example the button press for triggering listening on the Mark-1 will never be triggered.

pmulh replied:
Oh yes, good point, I think you're right that it should have a is_playing check in there.

forslund commented:
Hey I think this is a valid use case and I'd love to merge it. I think your changes looks reasonable and will do quick verification to make sure the default case is functional. After that I will merge since this looks fairly good. There may be issues like the comment I made on the stop signal. Fixing it can wait since it shouldn't affect the normal control.

forslund commented:
Seems to be working fine, will be merging it in it's current state. It would be great if you can do a follow up PR with some sort of addition to the README.md to explain the feature.

pmulh commented:
> Seems to be working fine, will be merging it in it's current state. It would be great if you can do a follow up PR with some sort of addition to the README.md to explain the feature.

Great, thanks for merging. Yes, I'll do a follow-up with an explanation and a fix for that issue on the stop signal you mentioned.
